First version of VRE integrated chart

volodymyrss opened the pull request:
*  REANA
  *  NFS provisioner
* jupyterhub with configuration for rucio access
  *  Rucio (?)

paullaycock commented on vre/templates/reana-ingress.yaml:
```diff
@@ -0,0 +1,45 @@
+    traefik.frontend.entryPoints: http,https
+  name: {{ .Release.Name }}-reana-ingress
+spec:
+  ingressClassName: traefik
```
```suggestion
  +  ingressClassName: {{ .Values.reana.ingress.ingressClassName }}
```

paullaycock commented on vre/values-custom.yaml:
```diff
@@ -0,0 +1,7 @@
+reana:
```
Wishlist to add
hostbase or similar, that is templated for the jhub and reana host names
Make indigo IAM client secret available for jhub too
Indigo IAM authorise, token and callback URLs for reana and jhub
nfs storage classes, both the default for persistent, and the reana shared storage class
common (postgres)db secret
Rucio host (template the extraEnv and rucio.cfg file generation ?)

volodymyrss replied:
Note that it is not that straightforward to remove all duplication from values. It is not possible to use template within values file itself.
It is possible to use yaml anchors but it's not for all cases (e.g. not when substring need to be templated).

Usually to avoid duplication like that, it is necessary to modify the chart itself, in this case the dependency charts.
I will see what can be done without being too invasive.

We could also template values ourselves, but it means adding a custom tool on top of helm and this will be less portable.

paullaycock commented on skaffold.yaml:
```diff
@@ -0,0 +1,13 @@
+deploy:
+  helm:
+    releases:
+      - name: cern-vre
```
maybe "escape-vre" is better, or even "my-escape-vre" as a default release name.  Whatever this thing is, it won't actually be the CERN VRE :)

volodymyrss replied:
I see, interesting question. 

For me, "VRE" is a generic term which I see throughout the related domains. So I assumed that name of the software product is "CERN VRE". 

It also coincides with the name of the CERN instance.

However, we have now ESCAPE VRE, CERN VRE, ETAP VRE, and I wonder.

@garciagenrique @Soap2G what do you think? Should "CERN" be anywhere in the name of the generic VRE chart?

volodymyrss replied:
Renamed to ESCAPE VRE.

paullaycock commented on vre/values.yaml:
```diff
@@ -0,0 +1,485 @@
+      blockWithIptables: false 
+
+    extraEnv:        
+      RUCIO_AUTH_HOST: "https://vre-rucio-auth.cern.ch"
```
I think this should move to a top-level configuration as eventually I think we will want to configure REANA to use rucio (with token auth)

paullaycock commented on vre/values.yaml:
```diff
@@ -0,0 +1,485 @@
+                echo -n $RUCIO_ACCESS_TOKEN > /tmp/rucio_oauth.token;
+                mkdir -pv /opt/rucio/etc;
+                echo "[client]" >> /opt/rucio/etc/rucio.cfg;
+                echo "rucio_host = $RUCIO_HOST" >> /opt/rucio/etc/rucio.cfg;
```
Define rucio host once

paullaycock commented on vre/values.yaml:
```diff
@@ -0,0 +1,485 @@
+
+  ingress:
+    enabled: true
+    ingressClassName: null
```
This is at least confusing to me, the annotations below assume traefik

volodymyrss replied:
there is a traefik built in reana

paullaycock commented on vre/values.yaml:
```diff
@@ -0,0 +1,485 @@
+
+  ingress_override: true
+
+  reana_hostname: "reana-vre.obsuks4.unige.ch"
```
template it so that there's a hostbase path or pick a name, and jhub and reana hostnames are derived from that?

volodymyrss commented on vre/values.yaml:
```diff
@@ -0,0 +1,485 @@
+        consumer_secret: "testsecret"
+
+    database:
+      user: 
```
This is to be set from secret in flux, or locally as needed. By default I'd put nothing. Maybe it can be generated like postgres chart does.

volodymyrss commented:
I added a [job](github.com/vre-hub/helm-charts/blob/first-version-vre-chart/vre/templates/bootstrap.yaml) which will run bootstrap commands, so it should install without manual interventions. I did not test much yet, but please do give it a try, @paullaycock .

paullaycock commented on vre/values.yaml:
```diff
@@ -0,0 +1,485 @@
+                echo "[client]" >> /opt/rucio/etc/rucio.cfg;
+                echo "rucio_host = $RUCIO_BASE_URL" >> /opt/rucio/etc/rucio.cfg;
+                echo "auth_host = $RUCIO_AUTH_URL" >> /opt/rucio/etc/rucio.cfg;
+                echo "account = $JUPYTERHUB_USER" >> /opt/rucio/etc/rucio.cfg;
```
At least when using the ESCAPE IAM + Rucio instance

paullaycock commented on vre/values-custom-example.yaml:
```diff
@@ -0,0 +1,99 @@
+# please revise and set all these values!
+
+bootstrap:
+  adminEmail: "admin@example.com"
```
reanaAdminEmail / Password might be better, just to be clear where they're used

volodymyrss commented:
There is a question right now if it makes sense to use try using new pre-release reana 0.95.0, or port the CRM changes back onto 0.9.4.

garciagenrique commented:
> There is a question right now if it makes sense to use try using new pre-release reana 0.95.0, or port the CRM changes back onto 0.9.4.

I think it is always a good idea to try the latest version @volodymyrss. Any reason to use 0.9.4 ?

volodymyrss commented:
> > There is a question right now if it makes sense to use try using new pre-release reana 0.95.0, or port the CRM changes back onto 0.9.4.
> 
> I think it is always a good idea to try the latest version @volodymyrss. Any reason to use 0.9.4 ?

The latest version is not released, it's in alpha. 

Right now I am having an issue with it, related to parsing user token info from IAM . I think it was working before though, so it's strange the issue appeared both in dev and staging setup. Maybe IAM changed. 

Anyway, alpha is not stable and will require continuous updates to our development.

When we last talked with @tiborsimko I think it was said it's too early to try 0.95.0, but I also did not see a huge number of changes in reana github org since, and the latest release is 0.95.0-alpha.2 is in Feb.
Do you know @tiborsimko what is the plan for 0.95.0, should we try to contribute our [monitoring changes](https://github.com/issues?q=state%3Aopen%20archived%3Afalse%20sort%3Aupdated-desc%20%20org%3Areanahub%20label%3A%22community%2Fetap%22)  to it?

Or @garciagenrique you probably know of what is the plan for REANA version adoption in the VRE?

volodymyrss commented:
I remembered why we moved to alpha version of REANA. 

It's partly because we wanted to use included DB, but I saw this https://github.com/reanahub/reana/blob/0.9.4/helm/reana/templates/reana-db.yaml#L45 which seems to set debug values only in non-debug environment - so there seems to be maybe a bug? Of course we could use external db but it seems strange.

Second reason is that we are actually developing reana chart, so it's weird to start from old version.

But then, current alpha version stopped working, so this needs some investigation.

edit: 

I think alpha version is not compatible with Indigo IAM, because recent versions of `invenio_oauthclient` seem to assume fetching userinfo both from endpoint and from token, specifically in the way keycloack does it. IAM fails with token userinfo decoding because it can not get the token signature right.

I am sure it is possible to adjust `invenio_oauthclient` and `reana_server` (which needs some extra configuration in this case).

In principle until 0.95.0 is stable, we could use 0.9.4 with a patch for the DB credentials in debug mode.
But 0.9.4 uses python3.8 which is already EoL.

Overall, it seems that to finish our project, we need a bit deeper changes to involvement reana and invenio_oauthclient, and count on new reana to be released.

An advice from @tiborsimko would be welcome, but I can also look myself deeper and propose further reana PRs.

volodymyrss commented:
We discussed we need to talk about storage usage. To avoid duplication of data.

We are also thinking about using RSE per VRE instance.

degauden commented:
Some remarks about the test deployment that I have done on https://reana-vre.obsuks5.unige.ch:
- I did the deployment in the test namespace
  ```shell
  $ helm install my-vre vre/escape-vre -n test --create-namespace --devel
  ```
  the `--devel` option is needed since there is no production chart yet.
- The REANA_ACCESS_TOKEN is in the log of the install. But it can be retrieved afterwards with
  ```shell
  $ export REANA_ACCESS_TOKEN=$(kubectl get secret escape-vre-admin-access-token -o json | jq -r '.data | map_values(@base64d) | .ADMIN_ACCESS_TOKEN')
  ```
- Some extra arguments are needed
    - `--set nfs-server-provisioner.persistence.storageClass=vspherecsi-resize --set nfs-server-provisioner.persistence.size=20Gi ` : for the NFS provisioning
    - `--set reana.reana_hostname=reana-vre.obsuks5.unige.ch` for the hostname
    - `--set jupyterhub.hub.config.RucioAuthenticator.oauth_callback_url=https://jhub-vre.obsuks5.unige.ch/hub/oauth_callback` for the Rucio callback url
    - the name of the release cannot by a custom `my-vre`, but has to be `escape-vre`
    - the scerets from INDIGO IAM have to be passed with ` --set reana.secrets.login.iam.consumer_key=XXXXXXXXXXXXXX --set reana.secrets.login.iam.consumer_secret=YYYYYYYYYYYYYYYY`

degauden commented:
Finally the install command looks like:
```shell
$ helm upgrade --install escape-vre vre/escape-vre -n test --create-namespace --devel \
 --set nfs-server-provisioner.persistence.storageClass=vspherecsi-resize \
 --set nfs-server-provisioner.persistence.size=20Gi --set reana.reana_hostname=reana-vre.obsuks5.unige.ch \
 --set jupyterhub.hub.config.RucioAuthenticator.oauth_callback_url=https://jhub-vre.obsuks5.unige.ch/hub/oauth_callback\
 --set reana.secrets.login.iam.consumer_key=XXXXX --set reana.secrets.login.iam.consumer_secret=YYYYYYYYYYYYYY
```
Remarks:
- The user has to be authorized manually. Moreover, the email notification to the admin is not working, and thus some communication needs to be done between the user and the admin.
- To list the users:
  ```shell
  $ kubectl exec -i -t deployment/escape-vre-server -- flask reana-admin\
   user-list --admin-access-token $REANA_ACCESS_TOKEN
  ```
  (please mind the `escape-vre-` prefix
-  To authorize (the user had to try to connect  beforehand):
  ```shell
  $ kubectl exec -i -t deployment/escape-vre-server -- flask reana-admin token-grant \
  --email user@example.com --admin-access-token $REANA_ACCESS_TOKEN
  ```
  
One last thing: the redirect URIs to be used in INDIGO IAM have to be:
https://jhub-vre.obsuks5.unige.ch/hub/oauth_callback
https://reana-vre.obsuks5.unige.ch/api/oauth/authorized/keycloak/ (please mind the trailing "/")
https://reana-vre.obsuks5.unige.ch/hub/oauth_callback

volodymyrss commented:
> Finally the install command looks like:
> 
> ```shell
> $ helm upgrade --install escape-vre vre/escape-vre -n test --create-namespace --devel \
>  --set nfs-server-provisioner.persistence.storageClass=vspherecsi-resize \
>  --set nfs-server-provisioner.persistence.size=20Gi --set reana.reana_hostname=reana-vre.obsuks5.unige.ch \
>  --set jupyterhub.hub.config.RucioAuthenticator.oauth_callback_url=https://jhub-vre.obsuks5.unige.ch/hub/oauth_callback\
>  --set reana.secrets.login.iam.consumer_key=XXXXX --set reana.secrets.login.iam.consumer_secret=YYYYYYYYYYYYYY
> ```
> 
> Remarks:
> 
>     * The user has to be authorized manually. Moreover, the email notification to the admin is not working, and thus some communication needs to be done between the user and the admin.
> 
>     * To list the users:
>       ```shell
>       $ kubectl exec -i -t deployment/escape-vre-server -- flask reana-admin\
>        user-list --admin-access-token $REANA_ACCESS_TOKEN
>       ```
>       
>       
>           
>             
>           
>       
>             
>           
>       
>           
>         
>       (please mind the `escape-vre-` prefix
> 
>     * To authorize (the user had to try to connect  beforehand):
> 
> 
> ```shell
> $ kubectl exec -i -t deployment/escape-vre-server -- flask reana-admin token-grant \
> --email user@example.com --admin-access-token $REANA_ACCESS_TOKEN
> ```
> 
> One last thing: the redirect URIs to be used in INDIGO IAM have to be: https://jhub-vre.obsuks5.unige.ch/hub/oauth_callback https://reana-vre.obsuks5.unige.ch/api/oauth/authorized/keycloak/ (please mind the trailing "/") https://reana-vre.obsuks5.unige.ch/hub/oauth_callback

Thanks a lot @degauden , I add this to the readme as a case deployment, specializing generic instructions.

Soap2G commented:
Hi @volodymyrss , what's the status of this PR? 
Do you have everything that you need in there (for now?
Would it be ready for a review?

volodymyrss commented:
> Hi @volodymyrss , what's the status of this PR? Do you have everything that you need in there (for now? Would it be ready for a review?

Since I removed draft status it has been!

volodymyrss commented:
> Hi @volodymyrss!
> 
> These updates look great, thank you very much! I think a rebase would help a lot, could we try to aggregate commits and also use the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) layout?

Ok, I will make couple of cosmetic changes and squash 

> 
> On top of that I'm approving, I only have a few minor questions (the review has been long enough already):
> 
>     1. Default admin password is `adminpassword` - users MUST change this; also I saw test IAM credentials in values (`testkey/testsecret`). Is there a way to make them dummy without being hardcoded? I'm also fine in leaving them there for the moment, if it's too much work.

There are different approaches to this, but ultimately dummy values have to come from somewhere, right? And in that way, they'd be hardcoded. They may be in values like now, or in the template default. I find it's move obvious when they are in values, so they are less likely to be forgotten.

Or I can make them empty, and make install fail with a clear message if they are not set. Perhaps this is best.

> 
>     2. `blockWithIptables: false` has a `TODO: revert` comment. Should we do that?
> 
>     3. I can see a dependency from `volodymyrsavchenko/reana-workflow-controller`; is there any update on an upstream version?

That was our experiments when we hoped to update reana with ETAP-specific changes. I will drop it.

volodymyrss commented:
* addressed TODOs
* made default values for credentials empty, and added chart validation which crashes install with error if they are not set
* squashed

volodymyrss commented:
Thanks all! That should do in the first approximation. It's also live in https://etap.obsuks7.unige.ch/
